### PR TITLE
[FLINK-25857] Add committer metrics to track the status of committables

### PR DESCRIPTION
**Pull request by pvary**
## What is the purpose of the change

This is the implementation of main part of [FLIP-371: Provide initialization context for Committer creation in TwoPhaseCommittingSink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-371%3A+Provide+initialization+context+for+Committer+creation+in+TwoPhaseCommittingSink).

Modifies the `TwoPhaseCommittingSink.Committer` creation by providing a `CommitterInitContext` for the `createCommitter` method.

The change is build on the top of #23541 so please review those there, in only check the 2nd commit here.

## Brief change log

API changes:
- `TwoPhaseCommittingSink` - new `createCommitter `method with default implementation and deprecation
- `CommitterInitContext` - new object for committer initialization and metrics
- `SinkCommitterMetricGroup` - new metrics group for committer related metrics

Implementation:
- `InternalSinkCommitterMetricGroup` - implementation for `SinkCommitterMetricGroup`
- `CommitterOperator` - to create the `CommitterInitContext` and `SinkCommitterMetricGroup`. And modify the initialization to  use the new `Committer` creation method, and create the metrics group
- Updating the metrics in the `CommitRequestImpl` and `CheckpointCommittableManagerImpl`
- Propagate the new metricGroup object to finally arrive to the `CommitRequestImpl` where we need it (`CommiterOperator` -> `CommittableCollector` -> `CheckpointCommittableManagerImpl` -> `SubtaskCommittableManager` -> `SubtaskCommittableManager` -> `CommitRequestImpl`) - This part of the change is somewhat hard to follow, and I open for suggestions to make it better
- SinkAdapterV1 changes to use the new Committer creation method, and metrics group (`SinkAdapterV1`, `GlobalCommitterSerializer`, `GlobalCommitterOperator`)
- Updating the tests to propagate the metric group
- Adding a new test case to `SinkV2MetricsITCase` to make sure that the metrics are propagated as expected.

## Verifying this change

Added a new unit test, and modified the old ones to use the new method.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented - might be worth to check the documentation of SinkV2 if it is exists, to update it.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommitRequestImpl.java [hunk 1]
```diff
@@ -32,16 +33,27 @@ public class CommitRequestImpl<CommT> implements Committer.CommitRequest<CommT>
...
 
-    protected CommitRequestImpl(CommT committable) {
+    protected CommitRequestImpl(CommT committable, SinkCommitterMetricGroup metricGroup) {
         this.committable = committable;
+        this.metricGroup = metricGroup;
         state = CommitRequestState.RECEIVED;
+
+        // Currently only the SubtaskCommittableManager uses this constructor to create a new
+        // CommitRequestImpl, so we can increment the metrics here
+        metricGroup.getNumCommittablesTotalCounter().inc();
```
Comment (tzulitai) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommitRequestImpl.java:45]: this feels a bit hacky, as your comment already hints. My main issue with this is that this happening in the constructor, if we only look at this class locally it's hard to tell if we're incrementing it correctly. For example, a deserializer for `CommitRequestImpl`s can totally call this constructor and unintentionally increment the metric.

I see that you've already wired in the `SinkCommitterMetricGroup` to the `SubtaskCommittableManager`. Can we just increment the total # of committables there in the `add` method then?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java [hunk 2]
```diff
@@ -72,7 +86,7 @@ void add(CommittableWithLineage<CommT> committable) {
 
     void add(CommT committable) {
         checkState(requests.size() < numExpectedCommittables, "Already received all committables.");
-        requests.add(new CommitRequestImpl<>(committable));
+        requests.add(new CommitRequestImpl<>(committable, metricGroup));
```
Comment (tzulitai) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java:89]: Increment total # of committables here instead of within `CommitRequestImpl` constructor.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkV2MetricsITCase.java [hunk 3]
```diff
@@ -116,6 +125,58 @@ public void testMetrics() throws Exception {
         jobClient.getJobExecutionResult().get();
     }
 
+    @Test
+    public void testCommitterMetrics() throws Exception {
```
Comment (tzulitai) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/SinkV2MetricsITCase.java:129]: 👍

**Review by tzulitai — COMMENTED** [review]
Thanks @pvary!

I've left a few comments on a first pass through the second commit.

Apart from those comments, I'd like to do a second pass on the following:
it looks like wiring in the new new committables metrics groups is adding quite a bit of "fat" w.r.t. external dependencies of various classes, e.g. `CommittableCollector` and the various serializers. I say they are redundant "fat" because ultimately only the end of the dependency chain, i.e. `CheckpointCommittableManagerImpl`, `SubtaskCommittableManager`, and `CommitRequestImpl` is doing meaningful logic on the metric group - all others earlier in the chain are just passing down the metric group and it looks rather polluted. This is rather evident in the test code changes, and isn't so nice.

This is really just a comment on cosmetics and arguably trivial, but I think we should spend a bit of time taking a stab to see if there's a better way around this. Lets timebox a day on this - I'm fine if we don't figure out a better path and leave it as is.

**Comment by pvary** [comment]
> Apart from those comments, I'd like to do a second pass on the following: it looks like wiring in the new new committables metrics groups is adding quite a bit of "fat" w.r.t. external dependencies of various classes, e.g. `CommittableCollector` and the various serializers. I say they are redundant "fat" because ultimately only the end of the dependency chain, i.e. `CheckpointCommittableManagerImpl`, `SubtaskCommittableManager`, and `CommitRequestImpl` is doing meaningful logic on the metric group - all others earlier in the chain are just passing down the metric group and it looks rather polluted. This is rather evident in the test code changes, and isn't so nice.

I tried before posting the PR, but were not able to come up with a better solution. 

After your review I did another stab on it:
https://github.com/pvary/flink/tree/committable_context,  https://github.com/pvary/flink/commit/7fce697e8ff586e80d810eed2c0cf48eb66c9f1b

What I did here, is created a `CommittableContext` with `checkpointId`, `subtaskId`, `metricsGroup`, and propagated this context instead of the individual parameters. This causes a behavioural change - the `null` checkpointId has been changed to `Long.MAX_VALUE`, which I think is not acceptable. See: https://github.com/pvary/flink/commit/7fce697e8ff586e80d810eed2c0cf48eb66c9f1b

Also the proposed change above only simplifies `CheckpointCommittableManagerImpl`, `CommitRequestImpl`, `SubtaskCommittableManager.java`, but does not help in `CommiterOperator` (uses/propagates only metricGroup), `CommittableCollector` (uses/propagates only metricGroup and subtaskId), `CommittableCollectorSerializer` (uses/propagates only metricGroup and subtaskId)

Sadly, I do not see any better options ATM 😢

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/TwoPhaseCommittingSink.java [hunk 4]
```diff
@@ -77,4 +97,42 @@ interface PrecommittingSinkWriter<InputT, CommT> extends SinkWriter<InputT> {
          */
         Collection<CommT> prepareCommit() throws IOException, InterruptedException;
     }
+
+    /** The interface exposes some runtime info for creating a {@link Committer}. */
+    @PublicEvolving
+    interface CommitterInitContext {
```
[the remark this answers is not in the document]
Comment (pvary) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/TwoPhaseCommittingSink.java]: Created a common `InitContext`. See: https://github.com/apache/flink/pull/23555/files#diff-d80129086eb15764c75da260ebb116757f01485854277f399da30af99713eb57

This `InitContext` serves as a base for `SinkV2.InitContext`, and the `TwoPhaseCommittingSink.CommitterInitContext`.

I am undecided on the naming of these classes/interfaces:
- Option 1: Stick to the current API specification. See: the PR.
- Option 2: Rename the old `SinkV2.InitContext` to `SinkV2.WriterInitContext`
    - It is easier to understand/use in the long run
    - Breaks the user code on 1.18->1.19 migration

I would vote for Option2 so we are better off on the long urun, but not sure how flexible can we be with the API.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommitRequestImpl.java (hunk 1 above)
Comment (pvary) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommitRequestImpl.java]: Moved.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java (hunk 2 above)
Comment (pvary) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java:89]: Moved

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/TwoPhaseCommittingSink.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/TwoPhaseCommittingSink.java]: Is it possible to go with option 2 + deprecate the existing `createWriter(InitContext)` method and have a new one `createWriter(WriterInitContext)` + have a backwards compatible path for existing implementors? In that case we should create a JIRA to keep track of removing all these deprecations in the future.

**Comment by tzulitai** [comment]
I gave the refactoring a stab as well, and also can't come up with something better. Looking around in other parts of the code, this seems to be a common problem in various classes. A dependency framework would help here to reduce the boilerplate clutter, but in the past we decided against using such frameworks.

Lets leave that aspect of the PR as is, and thank you for attempting to address that and gave that a try!

**Comment by pvary** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
+1 on `WriterInitContext`. The reasons behind is simple, we have `CommitterInitContext` which is properly naming what kind of context is expected. We should add the same clarity on the writer side.

**Comment by pvary** [comment]
@flinkbot run azure

**Comment by pvary** [comment]
Handing the `InitContext` -> `WriterInitContext` migration is not viable, because of the diamond inheritance, I went forward with migrating without the deprecation.

As a result the users has to slightly modify their Sink code when migrating from Flink 1.18 to Flink 1.19, but the new code is cleaner, and will not need another iteration when we remove the deprecated methods.

@MartijnVisser, @tzulitai: Please help me with the review!

Thanks,
Peter

**Comment by pvary** [comment]
@mbalassi: Rebased, and got a green run

**Review by mbalassi — APPROVED** [review]
Thanks, @pvary. Looks good now, I acknowledge that we had to disable the tests for now. I created FLINK-33528 to track reenabling them - let us aim to remove this technical debt by the time we release 1.19.

File: flink-core/src/main/java/org/apache/flink/api/connector/sink2/Sink.java [hunk 5]
```diff
@@ -55,17 +53,11 @@ public interface Sink<InputT> extends Serializable {
...
     /** The interface exposes some runtime info for creating a {@link SinkWriter}. */
     @PublicEvolving
-    interface InitContext {
-        /**
-         * The first checkpoint id when an application is started and not recovered from a
-         * previously taken checkpoint or savepoint.
-         */
-        long INITIAL_CHECKPOINT_ID = 1;
-
+    interface WriterInitContext extends InitContext {
```
Comment (Jiabao-Sun) [inline flink-core/src/main/java/org/apache/flink/api/connector/sink2/Sink.java:60]: I'm afraid this change may cause some compatibility issues for connectors.
WDYT? @dannycranmer, @MartijnVisser

Comment (MartijnVisser) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/Sink.java:60]: That's why we had a FLIP and a vote @Jiabao-Sun

Comment (Jiabao-Sun) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/Sink.java:60]: Maybe I misunderstood, but I didn't see any proposals regarding changes to Sink and InitContext in the FLIP-371 document. 
Would introducing CommitterInitContext inheriting InitContext meet the requirements?

Comment (pvary) [reply flink-core/src/main/java/org/apache/flink/api/connector/sink2/Sink.java:60]: We discussed the possibility of `CommitterInitContext` inheriting from the `InitContext`, but that would have been a short term solution only and the resulting API would be inconsistent. That is why the decision was to have `WriterInitContext` and `CommitterInitContext`, and accept that the connectors has to update the code.

If you are interested in the Source related changes please follow FLIP-372 too, as there will be similar changes there.
